blockchain/blocks/src/tipset.rs

```diff
@@ -55,103 +42,104 @@ impl ser::Serialize for TipSetKeys {
     /// specify identical epoch, parents, weight, height, state root, receipt root;
     /// contentID for headers are supposed to be distinct but until encoding is added will be equal.
     pub fn new(headers: Vec<BlockHeader>) -> Result<Self, Error> {
         // check header is non-empty
-        if headers.is_empty() {
-            return Err(Error::NoBlocks);
-        }
-
-        let mut sorted_headers = Vec::new();
-        let mut cids = Vec::new();
+        let (first_header, other_headers) = match headers.split_first() {
+            Some(x) => x,
+            None => return Err(Error::NoBlocks),
+        };
+
+        let verify = |predicate: bool, message: &'static str| {
+            if predicate {
+                Ok(())
+            } else {
+                Err(Error::InvalidTipset(message.to_string()))
+            }
+        };
 
         // loop through headers and validate conditions against 0th header
-        for i in 0..headers.len() {
-            if i > 0 {
-                // Skip redundant check
-                // check parent cids are equal
-                if !headers[i].parents().equals(headers[0].parents()) {
-                    return Err(Error::InvalidTipSet(
-                        "parent cids are not equal".to_string(),
-                    ));
-                }
-                // check weights are equal
-                if headers[i].weight() != headers[0].weight() {
-                    return Err(Error::InvalidTipSet("weights are not equal".to_string()));
-                }
-                // check state_roots are equal
-                if headers[i].state_root() != headers[0].state_root() {
-                    return Err(Error::InvalidTipSet(
-                        "state_roots are not equal".to_string(),
-                    ));
-                }
-                // check epochs are equal
-                if headers[i].epoch() != headers[0].epoch() {
-                    return Err(Error::InvalidTipSet("epochs are not equal".to_string()));
-                }
-                // check message_receipts are equal
-                if headers[i].message_receipts() != headers[0].message_receipts() {
-                    return Err(Error::InvalidTipSet(
-                        "message_receipts are not equal".to_string(),
-                    ));
-                }
-                // check miner_addresses are distinct
-                if headers[i].miner_address() == headers[0].miner_address() {
-                    return Err(Error::InvalidTipSet(
-                        "miner_addresses are not distinct".to_string(),
-                    ));
-                }
-            }
-            // push headers into vec for sorting
-            sorted_headers.push(headers[i].clone());
-            // push header cid into vec for unique check (can be changed to hashset later)
-            cids.push(headers[i].cid().clone());
+        for header in other_headers {
+            // Skip redundant check
```

Comment:
Can remove this comment now, was just included before when it was skipping the 0 index

Reply:
Whoops, fell through the cracks